src/geeknn/ordination/_base.py

```diff
@@ -0,0 +1,282 @@
+
+    @classmethod
+    def from_ee_feature_collection(
+        cls, fc: ee.FeatureCollection, num_threads: int = -1, chunk_size: int = 5000
+    ):
+        """Create a client-side FeatureCollection from an ee.FeatureCollection
+        using multiple threads and chunking to avoid memory issues.
+        """
+        size = fc.size().getInfo()
+        info = fc.limit(0).getInfo()
+        chunks = [fc.toList(chunk_size, i) for i in range(0, size, chunk_size)]
+
+        with Parallel(n_jobs=num_threads, backend="threading") as p:
+            chunk_data = p(delayed(chunk.getInfo)() for chunk in chunks)
+
+        info.update({"features": list(chain.from_iterable(chunk_data))})
+        return cls(**info)
+
+
+class Raw:
+    def __init__(self, k=1, max_duplicates=None):
+        self.k = k
+        self.max_duplicates = max_duplicates if max_duplicates is not None else 5
+
+    def transform_image(self, X_image: ee.Image) -> ee.Image:
+        """No-op placeholder."""
+        return X_image
+
+    def transform_fc(self, fc: ee.FeatureCollection) -> ee.FeatureCollection:
+        """No-op placeholder."""
+        return fc
+
+    @property
+    def k_nearest(self):
+        """Return the number of requested nearest neighbors plus max_duplicates
+        in order to account for non-independent samples.
+        """
+        return self.k + self.max_duplicates
+
+    def get_ids(self, fc: FeatureCollection, id_field: str) -> list[Any]:
+        """Return the IDs from the feature collection."""
+        return list(map(int, fc.aggregate_array(id_field)))
+
+    def _set_classifier(
+        self,
+        *,
+        fc: ee.FeatureCollection,
+        class_property: ee.String,
+        input_properties: ee.List,
+    ) -> None:
+        """Set the EE minimumDistance classifier for this estimator."""
+        self.clf = ee.Classifier.minimumDistance(
+            metric="euclidean",
+            kNearest=self.k_nearest,
+        ).train(
+            features=fc,
+            classProperty=class_property,
+            inputProperties=input_properties,
+        )
+
+    def train(
+        self, *, fc: ee.FeatureCollection, id_field: str, X_columns: list[str], **_
+    ):
+        """Train the estimator, storing the needed server-side objects used
+        in prediction.
+        """
+        self.id_field = ee.String(id_field)
+        self.X_columns = ee.List(X_columns)
+        self._set_classifier(
+            fc=fc, class_property=self.id_field, input_properties=self.X_columns
+        )
+        return self
+
+    def predict(self, X_image: ee.Image, mode: str = "CLASSIFICATION"):
+        """Predict the nearest neighbors for the given covariate (X) image."""
+        self.clf = self.clf.setOutputMode(ee.String(mode))
+
+        # Ensure the X_image band names match the X_columns
+        X_image = X_image.select(self.X_columns)
+
+        # Transform image
+        X_image = self.transform_image(X_image)
```

Comment:
Great idea!  With this latest commit, I've incorporated both this idea and your other idea for using `@singledispatchmethod` to make `Raw` and `Transformed` a bit more streamlined.  In order to make this refactoring, I think I have to create private methods (`Raw._predict_image` and `Raw._predict_fc`) as the setup before each of these calls is a bit different when using a `Raw` object versus a `Transformed` object.  For example, I created a new method (`Raw._get_ordered_X_image`) that passes the `X_image` and returns an image with possible reordering of the bands based on `self.X_columns`.  I couldn't put that into `Raw.predict` because a transformed image no longer has the same band names as the original (the same type of issues that we were dealing with when doing `sknnr` transformations).  

But your fix is a much better flow and more true to what I think of as good OOP practice as well as getting rid of the no-op functions!  Thanks for the good suggestion.

Reply:
>  In order to make this refactoring, I think I have to create private methods (Raw._predict_image and Raw._predict_fc) as the setup before each of these calls is a bit different when using a Raw object versus a Transformed object.

Ah, tricky. This looks like a good solution considering the constraints!